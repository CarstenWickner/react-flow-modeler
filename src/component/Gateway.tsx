import * as React from "react";

import { FlowElementWrapper } from "./FlowElementWrapper";
import { HorizontalStroke } from "./HorizontalStroke";

import { ConvergingGatewayNode, DivergingGatewayNode, ElementType, DivergingGatewayBranch } from "../types/ModelElement";
import { onLinkDropCallback } from "../types/EditAction";

const returnNull = (): null => null;

export class Gateway extends React.Component<{
    gateway: DivergingGatewayNode | ConvergingGatewayNode;
    onSelect: (reference: DivergingGatewayNode | DivergingGatewayBranch | ConvergingGatewayNode) => void;
    editMenu: (() => React.ReactNode) | undefined;
    onLinkDrop: onLinkDropCallback | undefined;
}> {
    onClick = (event: React.MouseEvent): void => {
        const { onSelect, gateway } = this.props;
        onSelect(gateway);
        event.stopPropagation();
    };

    render(): React.ReactNode {
        const { gateway, editMenu, onLinkDrop: onDrop, onSelect, children } = this.props;
        const cssType = gateway.type == ElementType.DivergingGatewayNode ? "diverging" : "converging";
        return (
            <>
                <FlowElementWrapper
                    elementTypeClassName={`gateway-element ${cssType}`}
                    referenceElement={gateway}
                    editMenu={editMenu}
                    onLinkDrop={onDrop}
                    onClick={this.onClick}
                />
                {gateway.type === ElementType.ConvergingGatewayNode && <div className="stroke-horizontal optional" />}
                {gateway.type === ElementType.DivergingGatewayNode && (
                    // if this gateway is selected, provide a function as editMenu to indicate that, but let the function not render a second menu
                    <HorizontalStroke referenceElement={gateway} editMenu={children && editMenu ? returnNull : undefined} onSelect={onSelect}>
                        {children}
                    </HorizontalStroke>
                )}
            </>
        );
    }
}
