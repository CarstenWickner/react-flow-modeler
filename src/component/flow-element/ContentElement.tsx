import * as React from "react";

export const ContentElement: React.FunctionComponent<{
    children: React.ReactChild;
}> = ({ children }) => (
    <>
        <div className="arrow" />
        <div className="flow-element content-element">{children}</div>
    </>
);
