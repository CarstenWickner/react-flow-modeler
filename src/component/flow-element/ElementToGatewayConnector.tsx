import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export const ElementToGatewayConnector: React.FunctionComponent<{
    connectionType: "first" | "middle" | "last";
}> = ({ connectionType }) => <HorizontalStroke outgoingConnection={connectionType} />;
