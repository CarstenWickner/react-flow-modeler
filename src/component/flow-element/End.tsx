import * as React from "react";

export const End: React.FunctionComponent = () => (
    <>
        <div className="arrow" />
        <div className="flow-element end-element" />
    </>
);
