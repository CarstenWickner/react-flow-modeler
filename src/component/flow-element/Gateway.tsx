import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export const Gateway: React.FunctionComponent<{
    children?: React.ReactChild;
}> = ({ children }) => (
    <>
        <div className="arrow" />
        <div className="flow-element gateway-element" />
        <HorizontalStroke>{children}</HorizontalStroke>
    </>
);
