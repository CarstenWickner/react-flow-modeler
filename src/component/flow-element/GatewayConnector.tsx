import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export const GatewayConnector: React.FunctionComponent<{
    incomingConnection: "first" | "middle" | "last";
    children?: React.ReactChild;
}> = ({ incomingConnection, children }) => <HorizontalStroke incomingConnection={incomingConnection}>{children}</HorizontalStroke>;
