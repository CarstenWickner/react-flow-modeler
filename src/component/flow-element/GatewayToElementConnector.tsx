import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export const GatewayToElementConnector: React.FunctionComponent<{
    connectionType: "first" | "middle" | "last";
    children?: React.ReactChild;
}> = ({ connectionType, children }) => <HorizontalStroke incomingConnection={connectionType}>{children}</HorizontalStroke>;
