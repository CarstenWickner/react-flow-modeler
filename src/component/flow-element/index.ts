export { Start } from "./Start";
export { ContentElement } from "./ContentElement";
export { ElementToGatewayConnector } from "./ElementToGatewayConnector";
export { Gateway } from "./Gateway";
export { GatewayToElementConnector } from "./GatewayToElementConnector";
export { StrokeExtension } from "./StrokeExtension";
export { End } from "./End";
