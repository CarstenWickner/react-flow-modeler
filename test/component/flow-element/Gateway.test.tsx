import * as React from "react";
import { shallow } from "enzyme";

import { Gateway } from "../../../src/component/flow-element/Gateway";

describe("renders correctly", () => {
    it("without props", () => {
        const component = shallow(<Gateway>text</Gateway>);
        expect(component).toMatchSnapshot();
    });
});
