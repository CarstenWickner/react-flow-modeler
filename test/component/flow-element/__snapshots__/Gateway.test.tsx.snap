// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly without props 1`] = `
<Fragment>
  <div
    className="arrow"
  />
  <div
    className="flow-element gateway-element"
  />
  <HorizontalStroke
    incomingConnection="single"
    outgoingConnection="single"
  >
    text
  </HorizontalStroke>
</Fragment>
`;
